test/runtime/futex.c

```diff
@@ -0,0 +1,401 @@
+    that they've been woken up */
+    for(int i = 0; i < num_to_wake; i++) {
+        if(pthread_join(threads[i], NULL) != 0) {
+            printf("Unable to join thread.\n");
+            return false;
+        }
+    }
+    if (ret == expected_result)
+        return true; /* success */
+    return false; /* error */
+}
+
+/* Thread function called from futex_wake_test
+which sets wake_test_futex's value to 1 and blocks */
+static void *futex_wake_test_thread(void *arg) 
+{
+    int new_val = 1;
+    wake_test_futex = new_val;
+    syscall(SYS_futex, (int*)(arg), FUTEX_WAIT, new_val, 0, NULL, 0);
+    return NULL;
+}
+
+/* FUTEX_WAIT test 1: Check for error -1 because
+value at uaddr doesn't match val */
+static boolean futex_wait_test_1() 
+{
+    int *uaddr = (int*)(&wait_test_futex);
+    int val = 20;
+    int expected_result = -1; /* error */
+    int ret = syscall(SYS_futex, uaddr, FUTEX_WAIT, val, 0, NULL, 0);
+    if (ret == expected_result) {
+        printf("wait test 1: passed\n");
+        return true;
+    }
+    printf("wait test 1: failed\n");
+    return false;
+}
+
+/* FUTEX_WAIT test 2: Success case where value at
+uaddr matches val. Check that the call waits for
+the given amount of time and then times out */
+static boolean futex_wait_test_2() 
+{
+    int *uaddr = (int*)(&wait_test_futex);
+    int val = FUTEX_INITIALIZER; /* value at uaddr */
+    int expected_result = -1;
+
+    /* timeout of 0.3 sec */
+    struct timespec timeout = {.tv_sec = 0, .tv_nsec = 300000000};
+    int ret = syscall(SYS_futex, uaddr, FUTEX_WAIT, val, &timeout, NULL, 0);
+    
+    /* Check timeout by checking value of errno */
+    if ((errno == ETIMEDOUT) && (ret == expected_result)) {
+        printf("wait test 2: passed\n");
+        return true;
+    }
+    printf("wait test 2: failed\n");
+    return false;
+}
+
+/* FUTEX_WAIT_BITSET test 1: Check for error -1
+because the value at uaddr and val do not match */
+static boolean futex_wait_bitset_test_1() 
+{
+    int *uaddr = (int*)(&wait_bitset_test_futex);
+    int val = 20;
+    int bitset = 0xffffffff;
+    int expected_result = -1; 
+
+    int ret = syscall(SYS_futex, uaddr, FUTEX_WAIT_BITSET, val, 0, NULL, bitset);
+    if (ret == expected_result) {
+        printf("wait_bitset test 1: passed\n");
+        return true;
+    }
+    printf("wait_bitset test 1: failed\n");
+    return false;
+}
+
+/* FUTEX_WAIT_BITSET test 2: Success case where
+the value at uaddr and val do match. Check that the 
+call waits for the given amount of time 
+and then times out */
+static boolean futex_wait_bitset_test_2() 
+{
+    int *uaddr = (int*)(&wait_bitset_test_futex);
+    int val = FUTEX_INITIALIZER; /* value at uaddr */
+    int bitset = 0xffffffff;
+    int expected_result = -1; 
+
+    /* timeout of 0.3 seconds */
+    struct timespec timeout = {.tv_sec = 0, .tv_nsec = 300000000};
+    int ret = syscall(SYS_futex, uaddr, FUTEX_WAIT_BITSET, val, &timeout, NULL, bitset);
```

Comment:
When FUTEX_WAIT_BITSET is given a timeout, this timeout is interpreted (unlike in the FUTEX_WAIT case) as an absolute time. So here the timeout value should be at an offset (e.g. 0.3 seconds) from the time retrieved via clock_gettime(CLOCK_MONOTONIC), as you were doing before.

Reply:
@francescolavra, that makes sense. I updated that test to make the timeout value an offset, as I had in a previous version of the code. Does it look correct now?

Reply:
The value in the tv_nsec field of struct timespec should always be in the range 0 to 999999999, and the current code doesn't ensure this.

Reply:
@francescolavra I've now added a check to make sure the timespec tv_nsec field is always between [0, 999999999]. Does it look correct now?

Reply:
The result of the division between integer variables is always an integer, so there is no need to call `floor()`, and the math.h header file inclusion can be avoided as well.

Reply:
@francescolavra I have fixed this so that floor is not called, and I removed the math.h in the header. Do all the futex tests look good now or is there anything I should update?